lib/client.js

```diff
@@ -139,7 +176,25 @@ module.exports = function (dependencies) {
           let error = new VError("stream ended unexpectedly");
           resolve({ device, error });
         }
-      })
+      });
+
+      request.setTimeout(this.config.requestTimeout, () => {
+        if (logger.enabled) {
+          logger('Request timeout');
+        }
+
+        request.close(NGHTTP2_CANCEL);
```

Comment:
Closing the request sends a message indicating that the **request** should be cancelled by the http2 library over the session, but does not close the session itself - new requests will continue to be sent over the session, even if the remote server is no longer sending any response to packets such as acknowledgements or indicating that the connection is invalid/terminated

https://en.wikipedia.org/wiki/HTTP/2#Goals - i.e. the `session` is a TCP socket, and the `request`s are multiplexed by the http2 library over a single TCP session.

Reply:
I know - I'm confused about what session.setTimeout is actually doing.

E.g. What happens if I repeatedly call setTimeout every 20 milliseconds - It appears at a glance as if the old timeout gets cleared and replaced with the new timeout, but I may be mistaken.
- If the old timeout does actually get cleared, that would be a problem if send() keeps getting called

Reply:
I believe we should go with the request.setTimeout. In my tests I've seen a request timing out but the next one in the same session successfully completed. I believe that if something happen to fail the whole session, we will have the error event being fired and the connection being renewed.

Reply:
@TysonAndre I tested calling session.setTimeout repeatedly (see : https://github.com/petitchevalroux/node/blob/http2-timeout/http2-timeout/session-timeout.js). I get the following output : 
```
session received
Wed Oct 07 2020 17:09:19 GMT+0200 (GMT+02:00)
session timeout 1
Wed Oct 07 2020 17:09:24 GMT+0200 (GMT+02:00)
session timeout 2
```
As there is no multiple "session timeout X" between date log, it seems to me that you are right saying that the old timeout is replaced every time. 

[Here](https://github.com/petitchevalroux/node/blob/http2-timeout/http2-timeout/request-timeout.js) I tested successfully (as @davimacedo) if a request can suceeed after a previous one timed out using the same session.

To me session.setTimeout handler is just trigger if nothing happen on the session (as the doc says). We could use it as a new check when sending a request but actually it does not add much compared to the setTimeout of the request. If the tcp socket time out, the request will too.

Sorry, I am now convinced that it is not necessary.

Reply:
> Here I tested successfully (as @davimacedo) if a request can suceeed after a previous one timed out using the same session.

You should be aware that https://nodejs.org/api/http2.html#http2_request_settimeout_msecs_callback has a different behavior if a callback or an event handler is declared for 'timeout' - It destroys the Http2Stream if there is no callback/event handler.

To actually destroy the request, you may need to call request.destroy - I think it might slowly leak memory otherwise, depending on what APNs is doing, but I'm not familiar with the implementation

> Each instance of the Http2Stream class represents a bidirectional HTTP/2 communications stream over an Http2Session instance. Any single Http2Session may have up to `2^31-1` Http2Stream instances over its lifetime.



> If no 'timeout' listener is added to the request, the response, or the server, then Http2Streams are destroyed when they time out. If a handler is assigned to the request, the response, or the server's 'timeout' events, timed out sockets must be handled explicitly.

Reply:
Anyway, it's possible that this might fix this - I wouldn't know until we try that, and your test results seem promising if the connection is still working.

It seems like https://developers.google.com/web/fundamentals/performance/http2#flow_control might help - I was initially concerned that the  NGHTTP2_CANCEL frame might have just been more data sent over the TCP socket, and if the tcp send buffer was full, APNs wouldn't get the frame, but if there's additional flow control, cancellation might work properly assuming it's more urgent than new requests.

I wonder if the send issues might be specific to OS versions or node versions or the hardware we're running on, but hopefully it's a reproducible bugs (e.g. only some users run into that and others don't)

Reply:
I actually used the same code that I found in node.js docs and that's why I believe it has more chances to be the right way to go: https://nodejs.org/api/http2.html#http2_http2stream_settimeout_msecs_callback

It successfully worked in my tests and I believe with should go with this and maybe change later if some new issues arise.